### DISCUSSION TITLE
Click Responses

### DISCUSSION THREAD

**Discussion by BobRoss048** [jspsych/jsPsych#1114] [Q&A] [answered]
Hello,

I was wondering if there was a way to "code" click responses in JsPsych?  

For example: I have an image that appears at the top of the computer screen & I want to code the top-half of that image as one response & the bottom half of that image as another response.  

Is there a way to code that in with the picture?  For my particular experiment, depending on the condition the participant(s) are in, the image is going to either be located at the top of the computer screen, or the bottom of the computer screen.

**Comment by becky-gilbert** [comment, marked as answer]
Hi @BobRoss048, yep this should be possible. If I were you I'd set this up by modifying one the image plugins, like image-button-response. You can just remove the code that adds the button choice(s) to the screen.

What you can do is [attach a 'click' event listener to the image](https://www.w3schools.com/js/js_htmldom_eventlistener.asp), and inside that function you can (1) determine the mouse x/y coordinates relative to the image, (2) categorize the response as being top or bottom, (3) save the coordinates/category to the trial data, (4) finish up the trial (cancel the event listener, clear the screen, save the data with jsPsych.finishTrial).

Some of this stuff is already in other plugin files, so you can use other plugins as a reference. The key thing will be figuring out the x/y coordinates relative to the image, because I think the event information that you get after a click event just contains the x/y coordinates relative to the whole screen. To get this to work on different screen sizes, and hence different image locations, you'll need to first figure out the offset between the top left corner of the image and the top left of the page, then you can use that offset to get the x/y coordinates relative to the image. You can see an example in [this stackoverflow comment](https://stackoverflow.com/a/58568016). Feel free to post your code if you get stuck!

**Reply by BobRoss048** [reply]
Heya @becky-gilbert ,

I apologize, I'm stuck.  I've only started using JavaScript/HTML & JsPsych at the start of August, so my understanding of basic code is lacking significantly compared to most on here!  

Down below is my code:  Would I be placing the event listener function in the "Image Trial" section, under "stimulus"?  I know you mentioned replacing the button choice option, so I did that.  However, would the event listener just go there as a replacement for an option?    



```
// create an empty timeline array
var timeline = [];
    // create list of prime words to choose from
    var prime_word_list = ["Upward", "Downward",];
    // shuffle the prime word order
    var prime_word_list_rand = jsPsych.randomization.sampleWithoutReplacement(prime_word_list);
    // set up other trial information
    var trial_info = [
        {image: 'img/Test2.jpg'},
        {image: 'img/Test3.jpg'},
        {image: 'img/Goose.jpg'},
        {image: 'img/hawk.jpg'}
    ];
    // for each object in the trial info array, add the corresponding prime word from the randomized list
    for (var i=0; i<trial_info.length; i++) {
        trial_info[i].prime_word = prime_word_list_rand[i];
    }
    trial_info = jsPsych.randomization.sampleWithoutReplacement(trial_info);
    trial_info = trial_info.slice(0,1);
    var prime_trial = {
        type: 'html-keyboard-response',
        stimulus: jsPsych.timelineVariable('prime_word'),
        choices: jsPsych.NO_KEYS,
        trial_duration: 500,
        response_ends_trial: true
    };
    var image_trial = {
        type: 'image-keyboard-response',
        stimulus: jsPsych.timelineVariable('image'),
        response_ends_trial: true
    };
    var trial_procedure = {
        timeline: [prime_trial, image_trial],
        timeline_variables: trial_info,
        response_ends_trial: true
    };



```

**Reply by becky-gilbert** [reply]
@BobRoss048 no need to apologize! There are tons of jsPsych users with very little web programming experience. This is exactly what the forum is for 😃 

I was actually suggesting that you edit one of the image _plugin_ files, rather than your HTML file. I'm not sure if you've seen the Advanced jsPsych YouTube video from the Moving Research Online workshop, but there's a [section in that video about plugin files](https://youtu.be/LP7o0iAALik?t=5817), which might be helpful. Josh also made a separate [video tutorial about creating custom plugins](https://youtu.be/XQcsFwAmbiw).

I would suggest editing the "jspsych-image-button-response.js" file, and renaming it with a descriptive name like "jspsych-image-top-bottom-click.js". When you rename a plugin file, you'll also want to change the plugin name in the file, towards the top:
```javascript
jsPsych.plugins["image-top-bottom-click"] = (function() {

  var plugin = {};

  jsPsych.pluginAPI.registerPreload('image-top-bottom-click', 'stimulus', 'image');
```
Next, inside `plugin.info`, I would comment out the `choices` parameter, because you won't need it for your plugin, and because the default value is `undefined`, which means that jsPsych will require that you pass a value for this parameter. (Another option is to just change the default value from `undefined` to something else).
```javascript
    //   choices: {
    //     type: jsPsych.plugins.parameterType.STRING,
    //     pretty_name: 'Choices',
    //     default: undefined,
    //     array: true,
    //     description: 'The labels for the buttons.'
    //   },
```
Then I would comment out all of the code between 'display buttons' (line 116) and 'show prompt if there is one' (line 137), because this code is adding the different button response choices to the HTML that will be added to the screen, and you don't need any button response choices.

Then I would comment out the `for` loop that adds event listeners to each button choice, since you don't have any buttons. This is around lines 147-152.

Now, instead of listening for a 'click' event on each of the response buttons, you instead want to listen to a 'click' event on the stimulus image itself. This function will receive a single argument that contains information about the click event, including the x/y coordinates. So this is where we can calculate the x/y coordinates relative to the image, save these values and categorize the response as top/bottom.
```javascript
display_element.querySelector("#jspsych-image-button-response-stimulus").addEventListener("click", function(e) {
    // code below taken from https://stackoverflow.com/questions/34867066/javascript-mouse-click-coordinates-for-image/58568016#58568016
    bounds = this.getBoundingClientRect();
    var left = bounds.left;
    var top = bounds.top;
    var x = e.pageX - left;
    var y = e.pageY - top;
    var image_height = e.srcElement.height;
    // this is a ternary operator that checks whether the y click position is less than half the image height,
    // if so then half = "top", otherwise half = "bottom"
    var half = y < image_height / 2 ? "top" : "bottom";
    console.log("half: ", half);
    var click_pos = {
        x: x,
        y: y,
        half: half,
    };
    after_response(click_pos);
});
```
The last thing this function does is call the `after_response` function that is defined in the plugin. The `after_response` function used to take the button choice that the participant made as its argument. This doesn't make sense for your plugin, so instead I'm passing the `click_pos` object so that it can be saved to the data.

Around line 159 there's a comment that says 'store response'. The code below this sets up the `response` object to store information about the response. You can get rid of the `response` key, since that was meant to store the button number that was clicked, and replace it with keys to store the x and y coordinates of the image click, as well as whether the click was on the top or bottom half:
```javascript
var response = {
  rt: null,
  click_pos_x: null,
  click_pos_y: null,
  click_half: null
};
```
Inside the `after_response` function, you'll want to remove the line that saves the response button choice to the data, and replace that with lines that save the x and y coordinates to the values that you set up in your `response` object above, e.g.
```javascript
response.click_pos_x = click_pos.x;
response.click_pos_y = click_pos.y;
```
You'll also want to comment out lines 178-182, because these lines are looking for the response buttons on the page and disabling them, so that the participant can't respond again by clicking another button. Again, these lines don't make sense since you have no response buttons. 

Instead, what you can do is remove the 'click' event listener from your image, so that only the first image click is registered. You can do that by adding this line:
```javascript
display_element.querySelector('#jspsych-image-button-response-stimulus').onclick = null;
```
Finally, inside the `end_trial` function, you'll want to remove the button response from the `trial_data` object, and add in the new information that you want to save:
```javascript
var trial_data = {
    rt: response.rt,
    stimulus: trial.stimulus,
    click_x: response.click_pos_x,
    click_y: response.click_pos_y,
    click_half: response.click_half,
};
```
I've attached the full plugin code as a txt file, so you'll just need to change the extension to .js. 
[jspsych-image-top-bottom-click.txt](https://github.com/jspsych/jsPsych/files/5381512/jspsych-image-top-bottom-click.txt)

Then you can test this plugin in your HTML file by reading the plugin file in a `<script>` tag in your `<head>` section, and replacing the `type` in your `image_trial` object so that it matches the plugin name:
```javascript
var image_trial = {
    type: 'image-keyboard-response',
    stimulus: jsPsych.timelineVariable('image'),
    response_ends_trial: true
};
```

**Reply by BobRoss048** [reply]
Hello @becky-gilbert 

First & Foremost, thank you so much for this entire breakdown of the code & explanation of how it all works.  It was extremely helpful & exactly what I was looking for.  I cannot thank you enough for your patience with me in helping figure this out.  

I do have a follow up question, if that's okay!  

```
display_element.querySelector("#jspsych-image-button-response-stimulus").addEventListener("click", function(e) {
    // code below taken from https://stackoverflow.com/questions/34867066/javascript-mouse-click-coordinates-for-image/58568016#58568016
    bounds = this.getBoundingClientRect();
    var left = bounds.left;
    var top = bounds.top;
    var x = e.pageX - left;
    var y = e.pageY - top;
    var image_height = e.srcElement.height;
    // this is a ternary operator that checks whether the y click position is less than half the image height,
    // if so then half = "top", otherwise half = "bottom"
    var half = y < image_height / 2 ? "top" : "bottom";
    console.log("half: ", half);
    var click_pos = {
        x: x,
        y: y,
        half: half,
    };
    after_response(click_pos);
});
```

Once I get the X & Y coordinates, you mentioned saving the values & categorizing them appropriately.  I was just curious, would I do that in this chunk of code above?  Or is that a different line of code all together.  

Thank you again for your time,

**Reply by becky-gilbert** [reply]
Yep, some of it is in that 'click' event listener, and the plugin should do that already. These are the lines in the 'click' event listener that are categorizing the response as top/bottom and keeping track of the data that should be saved:
```javascript
var half = y < image_height / 2 ? "top" : "bottom"; // if y click coordinate is less than half the image height, then the click was on the top half, otherwise the bottom half
console.log("half: ", half); // this line isn't necessary - it just prints out the value of half to the console to make it easier to test
var click_pos = {
  x: x,
  y: y,
  half: half,
};
after_response(click_pos);
```
The last line above passes the click x/y coordinates and 'half' categorization to the `after_response` function (which already existed in the plugin, because I was modifying the image-button-response plugin). In the `after_response` function, I added some lines to store the x/y coordinates and top/bottom category in the `response` object:
```javascript
response.click_pos_x = click_pos.x;
response.click_pos_y = click_pos.y;
response.click_half = click_pos.half;
```
Finally, in the `end_trial` function, everything that you want stored in the trial data is put together in the `trial_data` object. Again, this `trial_data` object already existed in the plugin that I was modifying, so I just had to add the x/y coordinates and top/bottom category into this object:
```javascript
var trial_data = {
  rt: response.rt,
  stimulus: trial.stimulus,
  click_x: response.click_pos_x,
  click_y: response.click_pos_y,
  click_half: response.click_half,
};
```
Then, the data is saved when the `trial_data` is passed to `jsPsych.finishTrial`:
```javascript
jsPsych.finishTrial(trial_data);
```
Sorry that I forgot to mention these other changes! But this should all be in the file that I attached, so if you want to go through it, you can see my changes by comparing that file to the image-button-response plugin file. I hope that helps!

**Comment by kurokida** [comment]
I don't fully understand what you want to do, but my plugin might help.

[jspsych-psychophysics](https://jspsychophysics.hes.kyushu-u.ac.jp/)
